### DISCUSSION TITLE
Allow URL param to change the value of a React Context

### DISCUSSION THREAD

**Discussion by khash** [vercel/next.js#41556] [Help] [answered]
Hi 👋🏼 

I am building a documentation site, with a dropdown for the framework used, like "NextJS", "React", "Rails", ... 
Changing the selected item on this dropdown, changes parts of the content, using a React Context. 

Now, I'd like to be able to allow the user to pass in the framework as a URL query parameter. If the parameter is available, the dropdown will change to the value and the content will reflect that, if the parameter is not present, it will show the default framework.

<img width="228" alt="Screen Shot 2022-10-19 at 12 53 19 PM" src="https://user-images.githubusercontent.com/1121822/196684185-d16fce47-eeaa-4fe1-aa07-5ae685fdb65f.png">

Here is my `_app.jsx`:

```js
const [framework, setFramework] = useState(Frameworks[0])

  return (
    <>
      <AppContext.Provider value={{
        state: {
          framework: framework,          
        },
        actions: {
          setFramework: setFramework,
        }}}>
        <Layout>
          <Component {...pageProps} />
        </Layout>
      </AppContext.Provider>
    </>
  )
```

A `FrameworkSelector` component, calls `setFramework` when the user chooses a different framework and all works well.

However, doing something like this doesn't work as `router.query.framework` seems to be always undefined:

```js
  const router = useRouter()
  const [framework, setFramework] = useState(router.query.framework || Frameworks[0])
```

This application is going to do server side rendering and I don't need generation of static pages for it.

**Comment by icyJoseph** [comment, marked as answer]
Hi

I think the issue here is that `router.query` is initially empty because of pre-rendering.

> query: Object - The query string parsed to an object, including [dynamic route](https://nextjs.org/docs/routing/dynamic-routes) parameters. It will be an empty object during prerendering if the page doesn't use [Server-side Rendering](https://nextjs.org/docs/basic-features/data-fetching/get-server-side-props). Defaults to {}

In short if a page has no data-fetching functions, and there's no `getInitialProps` in _app or, if it has getStaticProps regardless of `getInitialProps` presence in _app, the page is optimized, that is, pre-rendered at build time. 

Because at build time there's no request, the page is optimized with an empty query.

Later at runtime, the page is hydrated with an empty query and immediately after Next.js triggers an update with the query object. 

I think if you define, `getInitialProps` for `_app`, and never define `getStaticProps` on pages, you'll be in good shape... 

```ts
import type { AppProps } from "next/app";
import { useRouter } from "next/router";

function MyApp({ Component, pageProps }: AppProps) {
  console.log(useRouter().query);
  return <Component {...pageProps} />;
}

MyApp.getInitialProps = () => {
  return {};
};

export default MyApp;
```

Something like that... but I'd challenge your approach a bit by saying that you probably don't need `useState` and the context provider... because the Router is already using those:

```ts
const Frameworks = ["foo", "bar", "baz"];

export const useFramework = () => {
  const router = useRouter();
  const framework = router.query.framework; // perhaps you should validate this one too :) - short on time

  const setFramework = useCallback((next?: string) => {
    if (typeof next === "undefined" || Frameworks.some((f) => f === next)) {
      Router.push({
        pathname: Router.pathname,
        query: { ...Router.query, framework: next },
      });
    }
  }, []);

  useEffect(() => {
    if (router.isReady && typeof router.query.framework === "undefined") {
      setFramework(Frameworks[0]);
    }
  }, [router.isReady, router.query, setFramework]);

  return { framework, setFramework }; // or [framework, setFramework] as const
};
```

Regardless of the above, if there's no `getInitialProps` in `_app` or the page is optimized at build time, then there'll be a frame where framework is `undefined`.

If you do that `getInitialProps` trick in `_app`, and the page doesn't implement `getStaticProps`, then the `useFramework` hook:

- going to /?framework=foo, has framework will be present from the start
- going to /, has a frame without framework (in strict mode, there'll be two) and then framework is available

Try it out, and let me know how it feels.

**Comment by khash** [comment]
Thank you for the help @icyJoseph!

I ended up using `next-usequerystate` to put the state in the URL but also have to use a context for other reasons (this site uses Markdoc and the components defined in Markdoc will read the context).

One other thing I ended up doing was to wrap the Query State in a second state and `useEffect` to avoid first render hydration issues:

```js
  let [queryFramework, setFramework] = useQueryState('framework')
  let [framework, setState] = useState();
  useEffect(() => {
    if (queryFramework === null) {
      queryFramework = GetDefaultFramework()
    }

    setState(queryFramework)
    SetDefaultFramework(queryFramework)
  }, [queryFramework]);
```

`GetDefaultFramework` and `SetDefaultFramework` deal with `localStorage` to persist user's preferences.
